180 bind api get projects

GoetzDeBouville opened the pull request:
Провайдит сеть в [экран списка проектов профиля](https://www.figma.com/design/tNhKnPzlgQNt0VpI2zODzi/%D0%A1%D1%82%D0%B0%D1%80%D1%82%D0%B0%D0%BF?node-id=3826-2902&t=sTlK4ni5k8oACVMM-4)
Добавляет общие классы для обработки ошибок
корректирует модели
корректирует фрагмент и адаптер
переводит обзервер со stateflow на livedata (если что, на BaseViewModel юзается stateflow)
фиксит обработку ошибок
удаляет лишнюю навигацию
в BaseViewModel добавляет навигацию на экран авторизации с очисткой бэкстэка

GoetzDeBouville committed:
180_добавляет модели данных

GoetzDeBouville committed:
180_добавляет конвертер

GoetzDeBouville committed:
180_добавляет сервис запроса списка проектов

GoetzDeBouville committed:
180_добавляет примеры изображений ошибок и тексты ошибок

GoetzDeBouville committed:
180_добавляет вертстку макета с сообщением ошибки

GoetzDeBouville committed:
180_добавляет enum класс для статуса экрана ошибок
расширяет ErrorT

GoetzDeBouville committed:
180_добавляет ApiService для получения списка проектов юзера

GoetzDeBouville committed:
180_добавляет метод расширение для класса ошибок для конвертации Response в Resource

GoetzDeBouville committed:
180_добавляет репозиторий и usecase для провайда сервиса

GoetzDeBouville committed:
180_расширяет ProjectsScreenState и провайдит stateflow во viewmodel

GoetzDeBouville committed:
180_провайдит зависимсоти и корректирует модули экрана проектов юзера

GoetzDeBouville committed:
180_добавляет верстку экрана ошибки

GoetzDeBouville committed:
180_рефакторит адаптер recyclerview

GoetzDeBouville committed:
180_расширяет список ошибок
добавялет класс для обработки ошибок

GoetzDeBouville committed:
180_корректирует верстку и стили
добавляет расширение для проверки сети
добавляет методы в BaseViewModel

GoetzDeBouville committed:
180_добавляет get метод в ProfileDao
добавляет отдельный networkclient и провайдит в него api

GoetzDeBouville committed:
180_корректирует репозиторий и usecase
корректирует viewmodel

GoetzDeBouville committed:
180_првоайдит зависимости

GoetzDeBouville committed:
180_добавляет обработку ошибки UNAUTHORIZED
форматирует код
удаляет не используемый код

GoetzDeBouville committed:
180_убирает сериализацию из моделей

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/data/network/ProjectListNetworkClientImpl.kt:
```diff
@@ -0,0 +1,50 @@
+            } catch (exception: HttpException) {
+                Response().apply { resultCode = exception.code() }
+            }
+        }
```
Сейчас возвращает 400 Bad Request, после того как убрал параметры пагинации возвращало 200, возможно проблема на стороне бэка, при этом ошибку не показывает т.к. при любом ответе присваивается SUCCESS_CODE

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/data/dto/ProjectDto.kt:
Лучше проставлять у сетевых моделек, которые будут создаваться через GsonConverter, аннотацию SerializedName(). Это нужно чтобы GsonConverter мог работать дальше после затирания имен полей обфускацией

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/data/converter/ProjectConverter.kt:
```diff
@@ -0,0 +1,27 @@
+import com.innoprog.android.feature.projects.domain.models.Project
+import com.innoprog.android.feature.projects.domain.models.ProjectAttachment
+
+fun ProjectDto.convert() = Project(
```
К данному проекту не применимо, но лучше давать более явные названия методам конвертации . Очень часто в больших проектах кроме Domain модели есть еще ui модель.

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/data/repository/ProjectRepositoryImpl.kt:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private companion object {
+        const val ERROR_TAG = "ProjectRepository"
```
Лучше не хардкодить теги, имена классов могут изменить, а про тег забыть, часто замечаю в легаси эти несостыковки 
```kotlin
val ERROR_TAG = ProjectRepository::class.simpleName
```

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/data/network/ProjectListNetworkClientImpl.kt:
```diff
@@ -0,0 +1,50 @@
+    private val apiService: ProjectApiService,
+    private val context: Context,
+    private val roomDB: RoomDB
+) : ProjectListNetworkClient {
```
Лучше сделать провайд именно profileDao, чтобы не было доступа к сущностям которые не используются в этом классе

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/di/ProjectsFragmentModule.kt:
```diff
@@ -0,0 +1,43 @@
+    fun bindNetworkCliend(impl: ProjectListNetworkClientImpl): ProjectListNetworkClient
+
+    @Binds
+    fun bundRepository(impl: ProjectRepositoryImpl): ProjectRepository
```
Опечатки в именах

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/domain/impl/GetProjectListUseCaseImpl.kt:
```diff
@@ -0,0 +1,23 @@
+    override suspend fun execute(): Resource<List<Project>> {
+        return when (val resource = projectRepository.getProjectList()) {
+            is Resource.Success -> {
+                val projects = resource.data.map { it.convert() }
+                Resource.Success(projects)
```
domain слой не должен знать о data, тут это правило нарушается т.к. используется моделька Data слоя

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/db/ProfileDao.kt:
```diff
@@ -3,9 +3,13 @@ package com.innoprog.android.feature.profile.profiledetails.data.db
 import androidx.room.OnConflictStrategy
+import androidx.room.Query
 
 @Dao
 interface ProfileDao {
     @Insert(onConflict = OnConflictStrategy.REPLACE)
     suspend fun saveProfile(profile: ProfileEntity)
+
+    @Query("SELECT * FROM profile_table LIMIT 1")
+    suspend fun getProfile(): ProfileEntity
 }
```
Добавлю задачку по очистке бд при повторной авторизации под другим акком, пока что поведение может быть неправильным

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/presentation/adapter/ProjectsScreenAdapter.kt:
```diff
@@ -0,0 +1,34 @@
+
+    override fun onBindViewHolder(holder: ProjectsScreenViewHolder, position: Int) {
+        holder.bind(listItem[position])
+        holder.itemView.setOnClickListener {
+            onItemClickListener(listItem[position].id.toString())
+        }
```
Не применимо к текущим потенциальным размерам списка, но лучше сразу вешать clickListener в onCreateViewHolder, чтобы делать это только один раз при создании

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/presentation/ProjectsScreenViewModel.kt:
```diff
@@ -1,29 +1,52 @@
-                if (it.first != null) {
-                    _state.value = ProjectsScreenState.Content(it.first!!)
-                } else {
-                    _state.value = ProjectsScreenState.Empty
+            when (val result = getProjectListUseCase.execute()) {
+                is Resource.Success -> {
+                    _state.value = if (result.data.isEmpty()) {
+                        ProjectsScreenState.Empty
+                    } else {
+                        ProjectsScreenState.Content(result.data)
+                    }
+                }
+
+                is Resource.Error -> {
+                    _state.value = renderError(result.errorType)
                 }
             }
```
Лучше через следующую структуру
```kotlin
viewModelScope.launch(Dispatchers.IO) { // По умолчанию viewModelScope запускается на Main потоке
    runCatching {
      // run useCase
    }.onFailure {
        // Дополнительно обрабатываем ошибки, например они могли возникнуть при маппинге модели в useCase или если другие разработчики что то сломают
    }
}
```

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/presentation/ProjectsScreenFragment.kt:
```diff
@@ -1,72 +1,143 @@
-        adapter.notifyDataSetChanged()
+    private fun renderError(errorState: ErrorScreenState) = with(binding) {
+        if (errorState == ErrorScreenState.UNAUTHORIZED) {
+            val direction = ProjectsScreenFragmentDirections
+                .actionProjectsFragmentToAuthorizationFragment()
+            viewModel.navigateTo(direction)
```
Думаю тут лучше использовать findNavController.popBackStack с параметрами 
```kotlin
        findNavController().popBackStack(destinationId = R.id.some_id, inclusive = true)
```

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/presentation/ProjectsScreenState.kt:
```diff
@@ -1,8 +1,11 @@
     data object Empty : ProjectsScreenState
-    data class Content(val projects: List<ProjectScreenModel>) : ProjectsScreenState
+    data object Loading : ProjectsScreenState
+    data class Error(val errorType: ErrorScreenState) : ProjectsScreenState
+    data class Content(val projects: List<Project>) : ProjectsScreenState
```
Именно Data class тут необязательно, он генерит доп методы которые мы не используем. Большое количество Data классов замедляют сборку релизной версии

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/projects/projectsScreen/presentation/ProjectsScreenViewModel.kt:
```diff
@@ -1,29 +1,52 @@
+            ErrorType.NO_CONNECTION -> ProjectsScreenState.Error(ErrorScreenState.NO_INTERNET)
+            ErrorType.NOT_FOUND -> ProjectsScreenState.Error(ErrorScreenState.NOT_FOUND)
+            ErrorType.INTERNAL_SERVER_ERROR -> ProjectsScreenState.Error(ErrorScreenState.SERVER_ERROR)
+            ErrorType.UNAUTHORIZED -> ProjectsScreenState.Error(ErrorScreenState.SERVER_ERROR)
```
ErrorScreenState.UNAUTHORIZED

GoetzDeBouville committed:
180_добавляет сериализацию для полей моделей

GoetzDeBouville committed:
180_фиксит обработку ошибок
фиксит запрос (количество проектов на страницу)
провайдит dao интерфейс в клиент вместо roomdb

GoetzDeBouville committed:
180_detekt formating

GoetzDeBouville committed:
Merge branch 'refs/heads/develop' into 180_bind_api_get_projects

GoetzDeBouville committed:
180_корректировки по замечаниям

GoetzDeBouville committed:
180_провайдит стэйт loading

GoetzDeBouville committed:
180_переводит проект на livedata
в базовую ВМ добавляет метод для очистки бэкстэка и перехода на экран авторизации

GoetzDeBouville committed:
180_detekt formatting

GoetzDeBouville committed:
Merge branch 'refs/heads/develop' into 180_bind_api_get_projects

GoetzDeBouville committed:
180_удаляет не используемую навигацию

GoetzDeBouville committed:
180_удаляет импорты дата слоя из домэйн

Commit:
Merge branch 'develop' into 180_bind_api_get_projects

# Conflicts:
#	app/build.gradle.kts

Commit:
Remove unnecessary set clickListener in bind method, add exception messages, add catching potential mapping exceptions

Secur1ty-21 reviewed (commented):
Убрал случайно уставленный clickListener в методе bind, добавил обработку потенциально возможных исключений при маппинге модели проекта в Domain ( Например пришли битые данные с сервера )
